appendices/VK_EXT_present_timing.txt

```diff
@@ -0,0 +1,255 @@
+to determine surface timing properties until after at least one image
+has been presented.
+
+In such cases, the special value of zero allows the application to
+indicate that timing feedback is desired, but that no
+targetPresentTime nor idealPresentTime is requested.
+Later, once the application has obtained feedback, it can specify
+targetPresentTime and idealPresentTime.
+
+
+7) How long before an application’s request for new image duration is honored?
+
+*UNRESOLVED*: Apparently, changes to some vendors' display hardware settings do
+not take effect immediately.
+It is not clear what settings, and therefore, it is not clear how to
+address this issue.
+
+
+8) Do we have a query for the anticipated latency from present to feedback?
+
+*UNRESOLVED*: There is some amount of latency from when an application calls
+vkQueuePresentKHR to when the image is displayed to the user, to when feedback
+is available to the application on when the image was actually displayed to the
+user.
+The first time (from the call till the image is presented) generally doesn’t
+matter, because the application will likely be providing a targetPresentTime
+(i.e. the application may have some indication for how long this will be).
+However, the latency between targetPresentTime until feedback is available may
+be much long.
+For example, on Android on the 1st-generation Pixel phone (60Hz FRR display),
+the latency was approximately 5 refresh cycles (83.33ms).
+For higher-frequency displays, the latency may have a larger number of refresh
+cycles.
+
+Is there value in having a query for the application to know how long it may
+have to wait for feedback?
+Can such a query be reasonably answered by the driver?
+
+Is there other interesting information in this space that we may wish to
+capture?
+
+
+9) Do we have a query(s) about the number of VkPastPresentationTimingEXT
+structs to keep?
+
+*UNRESOLVED*: At the Montreal F2F, there was discussion about how much
+feedback that the driver needs to keep and/or how much feedback that
+the application needs to be able to query.
+
+The way that the LunarG cube demo (official WSI example code) used
+VK_GOOGLE_display_timing, and what is proposed for the new extension is that
+the application query what feedback is available during every render-present
+loop.
+If the application never skips querying for feedback, and always obtains
+whatever feedback is available, there doesn’t seem much need for such a
+query(s).
+What I saw with the cube demo on a Pixel phone was that most of the time, the
+application obtained feedback for 1 previous present.
+Occassionally, it would get 2 VkPastPresentationTimingEXT structs on time and
+then 0 the next (or vice versa).
+
+Perhaps, a video player application might present several images at once, and
+then later get feedback for several images at the same time.
+That would be the most-likely use-case that I can come up with for why a query
+might be useful.
+Is that compelling enough?
+
```

Comment:
In neuroscience/vision-science testing it is not unusual to have paradigms where an application may prerender dozens or hundreds of frames - as much as it can fit into VRAM, and then wants to present them in a fast "burst", ie. queuing up many of them at once, specifying a present duration or target present time for each of the images in the sequence. Think about trying to present a sequence of complex images at 240 fps or more without skipping a frame. A variaton of this is to render ahead at least a few frames, like with triple-buffering etc. to maximize parallelism between gpu and cpu when complex rendering at high fps has to push the system hard, to avoid or minimize dropped frames.

In any of these scenarios it would be a big advantage to be able to render/queue ahead a lot and only collect the presentation timestamps at the end of such a burst. Atm. Psychtoolbox-3 on Linux + X11 + GLX windowing system interface + OpenGL takes advantage of the GLX extension INTEL_swap_events to allow for that. Wayland allows for the same with the current presentation_time extension. So for research applications, having a deep feedback queue would be good to retain functionality we already have under Linux/X11 or Linux/Wayland.

Reply:
I think the size of the internal timing feedback queue should be defined by the application. As it may involve some non-trivial allocations from the implementation (such as internal synchronization primitives), my proposal would be to add a pNext structure extending VkSwapchainCreateInfoKHR and specify it there, e.g.:

```
     typedef struct VkSwapchainPresentTimingCreateInfoEXT {
         VkStructureType            sType;
         const void*                pNext;
         uint32_t                   timingQueueSize;
         // ...
     } VkSwapchainPresentTimingCreateInfoEXT;
```

The question then is how to specify the behavior when that queue is full of entries not yet queried by `vkGetPastPresentationTimingEXT`. From previous internal discussions, my understanding is that not all implementations would be able to support either a ring-buffer behavior (overwrite oldest entries) or easily drop new entries. Making that case invalid by adding a Valid Usage entry to vkQueuePresentKHR seems like a reasonable compromise.